File: ios/Cloudmoji/Cloudmoji/Views/Sleepy/BreathingCloud.swift
import SwiftUI

/// Where the cloud is in one breath.
///
/// Its own type rather than a `MascotMood` case, and that is a rule not a
/// preference: `MascotMood`'s raw values are the `mascot-<mood>` identifiers the
/// UI suites read the celebration out of, and `MascotMood.arbitrate` is product
/// law about which face outranks which. A sleeping face is neither — it is one
/// screen's animation state, and putting it in the shared enum would make every
/// exhaustive `switch` on mood in the app grow a case it can never see.
enum BreathPhase: String, Hashable {
    case inhale, hold, exhale
    /// The session is over. The cloud stops moving and the loop stops running.
    case asleep
}

/// The breathing itself: timings, easing, and where the scale is at time *t*.
///
/// Pure and static, because it is the part of this screen that can be silently
/// wrong. How the cloud *looks* asleep is a judgement for the eye; whether a
/// four-second inhale actually takes four seconds is arithmetic, and arithmetic
/// that drifted by a beat would still look plausible on a screenshot.
///
/// Timings are `reference/breathing-cloud.jsx` verbatim: 4s in, 2s hold, 6s out.
enum BreathingSession {
    static let inhale: TimeInterval = 4
    static let hold: TimeInterval = 2
    static let exhale: TimeInterval = 6
    static var cycle: TimeInterval { inhale + hold + exhale }

    /// `scale(0.75)` at the bottom of an exhale, `scale(1.1)` at the top of an
    /// inhale, and a little smaller than either once asleep.
    static let restScale: Double = 0.75
    static let peakScale: Double = 1.10
    static let asleepScale: Double = 0.72

    /// The durations a grown-up may pick, in minutes.
    static let choices: [Int] = [2, 5, 10]

    /// Where the cloud is at `t` seconds into a session of `duration` seconds.
    ///
    /// `duration <= 0` means "no session picked yet", and returns the resting
    /// pose rather than declaring the session instantly over — the picker draws
    /// a still cloud through this same function.
    static func state(at t: TimeInterval, duration: TimeInterval) -> (scale: Double, phase: BreathPhase) {
        guard duration > 0 else { return (restScale, .inhale) }
        guard t < duration else { return (asleepScale, .asleep) }

        // Negative time cannot happen — the caller measures forward from a start
        // date — but a clock that stepped backwards must not produce a scale of
        // NaN in front of a child who is trying to fall asleep.
        let p = max(0, t).truncatingRemainder(dividingBy: cycle)

        if p < inhale {
            return (restScale + eased(p / inhale) * (peakScale - restScale), .inhale)
        }
        if p < inhale + hold {
            return (peakScale, .hold)
        }
        let k = (p - inhale - hold) / exhale
        return (peakScale - eased(k) * (peakScale - restScale), .exhale)
    }

    /// The prototype's `0.5 - 0.5 * cos(pi * k)` — an ease-in-out over 0...1.
    /// Linear breathing reads as a machine; this is what makes it read as a
    /// chest.
    private static func eased(_ k: Double) -> Double {
        0.5 - 0.5 * cos(.pi * k)
    }

    /// How far through the session, 0...1. Drives both the dimming and the
    /// progress line.
    static func progress(at t: TimeInterval, duration: TimeInterval) -> Double {
        guard duration > 0 else { return 0 }
        return min(max(t / duration, 0), 1)
    }
}

/// The sleeping cloud.
///
/// Its own view rather than a fifth `MascotMood`, for the reason in
/// ``BreathPhase``. The body geometry is the same 120 × 78 transcription
/// `CloudMascot` uses — same circles, same rounded base, same underside shadow —
/// so the two clouds are recognisably the same character; only the face and the
/// glow differ.
struct BreathingCloud: View {
    let scale: Double
    let phase: BreathPhase
    /// 0 at the start of a session, 0.55 at the end. Everything that glows fades
    /// against it, so the screen gets quieter as the child does.
    var dim: Double = 0

    /// The web viewBox. Every coordinate below is in these units.
    private static let artWidth: CGFloat = 120
    private static let artHeight: CGFloat = 78

    /// Drawn at 200 × 130 in the prototype, which is this viewBox at 1.667×.
    static let renderedWidth: CGFloat = 200

    /// Sideways. A landscape phone gives about 400pt of height, and 130 of it
    /// for the cloud plus a title, three buttons and a caption overflowed —
    /// the cloud came out clipped against the top edge of the screen.
    static let compactRenderedWidth: CGFloat = 132

    /// How wide this cloud is actually drawn. Injected rather than read from the
    /// environment so the previews and the tests can size it directly.
    var width: CGFloat = BreathingCloud.renderedWidth

    private var isAsleep: Bool { phase == .asleep }
    /// Eyes close on the hold as well as in sleep — the pause at the top of a
    /// breath is where a drowsy face settles.
    private var eyesClosed: Bool { isAsleep || phase == .hold }

    var body: some View {
        art
            .frame(width: Self.artWidth, height: Self.artHeight)
            .scaleEffect(width / Self.artWidth)
            .frame(width: width, height: width * 0.65)
            // The breath itself. Applied outside the art's own scale so the two
            // do not fight: this one is driven every frame by `TimelineView`,
            // which is why it carries no `.animation` of its own.
            .scaleEffect(scale)
            // Flatten first, or the six overlapping cloud shapes each cast a
            // shadow onto their neighbours and the silhouette turns glassy —
            // the same trap `CloudMascot` documents.
            .compositingGroup()
            .shadow(color: Theme.moonlight.opacity(0.28 * (1 - dim)), radius: 14 * scale, y: 8)
            .accessibilityElement(children: .ignore)
            .accessibilityLabel("Cloudmoji")
            // Deliberately **not** `mascot-<mood>`: a second element answering to
            // that identifier would make every celebration lookup in
            // `WordsModeUITests` ambiguous.
            .accessibilityIdentifier("sleepy-cloud-\(phase.rawValue)")
    }

    private var art: some View {
        ZStack {
            halo
            cloudBody
            cloudHighlights
            undersideShadow
            blush
            eyes
            mouth
            if isAsleep { zzz }
        }
    }

    private var halo: some View {
        Ellipse()
            .fill(
                RadialGradient(
                    stops: [
                        .init(color: Theme.moonlight.opacity(0.5), location: 0),
                        .init(color: Theme.moonlight.opacity(0), location: 1),
                    ],
                    center: .center, startRadius: 0, endRadius: 54
                )
            )
            .frame(width: 108, height: 72)
            .position(x: 60, y: 46)
            .opacity(0.35 * (1 - dim))
    }

    private var cloudBody: some View {
        ZStack {
            circle(cx: 30, cy: 46, r: 20)
            circle(cx: 52, cy: 36, r: 23)
            circle(cx: 72, cy: 30, r: 26)
            circle(cx: 94, cy: 42, r: 19)
            circle(cx: 42, cy: 44, r: 16)
            RoundedRectangle(cornerRadius: 12)
                .frame(width: 96, height: 24)
                .position(x: 60, y: 60)
        }
        .foregroundStyle(Theme.cloudWhite)
    }

    private var cloudHighlights: some View {
        ZStack {
            circle(cx: 72, cy: 22, r: 12)
            circle(cx: 50, cy: 30, r: 8).opacity(0.7)
        }
        .foregroundStyle(Theme.cloudHighlight)
    }

    private var undersideShadow: some View {
        Ellipse()
            .frame(width: 88, height: 12)
            .position(x: 60, y: 68)
            .foregroundStyle(Theme.cloudShadow)
            .opacity(0.4)
    }

    private var blush: some View {
        ZStack {
            Ellipse().frame(width: 16, height: 9).position(x: 34, y: 58)
            Ellipse().frame(width: 16, height: 9).position(x: 86, y: 58)
        }
        .foregroundStyle(Theme.blush)
        .opacity(0.42)
    }

    @ViewBuilder private var eyes: some View {
        if eyesClosed {
            closedEyes
        } else {
            ZStack {
                Ellipse().frame(width: 4.8, height: 5.6).position(x: 46, y: 51)
                Ellipse().frame(width: 4.8, height: 5.6).position(x: 74, y: 51)
            }
            .foregroundStyle(Theme.eyes)
        }
    }

    /// `M40 52 Q46 47 52 52` and its mirror, as a path so it does not depend on
    /// a font shipping the arc glyph.
    private var closedEyes: some View {
        var path = Path()
        for cx in [CGFloat(46), CGFloat(74)] {
            path.move(to: CGPoint(x: cx - 6, y: 52))
            path.addQuadCurve(to: CGPoint(x: cx + 6, y: 52), control: CGPoint(x: cx, y: 42))
        }
        return path
            .stroke(Theme.eyes, style: StrokeStyle(lineWidth: 2.2, lineCap: .round))
            .frame(width: Self.artWidth, height: Self.artHeight)
    }

    @ViewBuilder private var mouth: some View {
        switch phase {
        case .asleep:
            Ellipse()
                .fill(Theme.blushBeaming.opacity(0.75))
                .frame(width: 8, height: 6)
                .position(x: 60, y: 62)
        case .inhale:
            // A small round mouth, the way a person taking a breath in has one.
            Ellipse()
                .fill(Theme.blushBeaming.opacity(0.7))
                .frame(width: 9, height: 8)
                .position(x: 60, y: 62)
        case .hold, .exhale:
            quad(from: CGPoint(x: 55, y: 61), to: CGPoint(x: 65, y: 61), control: CGPoint(x: 60, y: 69))
                .stroke(Theme.eyes.opacity(0.75), style: StrokeStyle(lineWidth: 1.6, lineCap: .round))
                .frame(width: Self.artWidth, height: Self.artHeight)
        }
    }

    private var zzz: some View {
        ZStack {
            Text("z")
                .font(Theme.body(12, .black))
                .foregroundStyle(Theme.moonlight.opacity(0.8))
                .position(x: 101, y: 14)
            Text("z")
                .font(Theme.body(9, .black))
                .foregroundStyle(Theme.moonlight.opacity(0.6))
                .position(x: 109, y: 5)
        }
    }

    private func circle(cx: CGFloat, cy: CGFloat, r: CGFloat) -> some View {
        Circle().frame(width: r * 2, height: r * 2).position(x: cx, y: cy)
    }

    private func quad(from: CGPoint, to: CGPoint, control: CGPoint) -> Path {
        var path = Path()
        path.move(to: from)
        path.addQuadCurve(to: to, control: control)
        return path
    }
}

#Preview("Breathing, all four phases") {
    ZStack {
        Theme.background.ignoresSafeArea()
        VStack {
            BreathingCloud(scale: 1.1, phase: .inhale)
            BreathingCloud(scale: 1.1, phase: .hold)
            BreathingCloud(scale: 0.75, phase: .exhale)
            BreathingCloud(scale: 0.72, phase: .asleep, dim: 0.55)
        }
    }
}

File: android/app/src/main/java/app/cloudmoji/android/ui/sleepy/BreathingCloud.kt
package app.cloudmoji.android.ui.sleepy

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.withTransform
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.cloudmoji.android.model.BreathPhase
import app.cloudmoji.android.ui.theme.Blush
import app.cloudmoji.android.ui.theme.BlushBeaming
import app.cloudmoji.android.ui.theme.CloudHighlight
import app.cloudmoji.android.ui.theme.CloudShadow
import app.cloudmoji.android.ui.theme.CloudWhite
import app.cloudmoji.android.ui.theme.MascotEyes
import app.cloudmoji.android.ui.theme.Moonlight

/**
 * The sleeping cloud. Ported from iOS `BreathingCloud`
 * (`Views/Sleepy/BreathingCloud.swift`).
 *
 * Its own composable rather than a fifth [app.cloudmoji.android.model.MascotMood]
 * — the same reason [BreathPhase]'s own doc gives: a sleeping face is one
 * screen's animation state, not product law about which mood outranks
 * which. The body geometry below is drawn from the same 120 x 78
 * coordinates [app.cloudmoji.android.ui.mascot.CloudMascot] uses — same
 * circles, same rounded base, same underside shadow — so the two clouds
 * read as the same character; only the face and the glow differ. The shapes
 * are duplicated here rather than shared with `CloudMascot.kt`'s private
 * draw functions, mirroring iOS's own choice: `BreathingCloud.swift` does
 * not call into `CloudMascot.swift` either — see that file's class doc.
 *
 * No ambient/cast shadow behind the art, matching
 * [app.cloudmoji.android.ui.mascot.CloudMascot]'s own documented gap: a
 * dynamic-radius glow tied to [scale] needs `RenderEffect`, API 31, and this
 * app's `minSdk` is 26.
 */
@Composable
fun BreathingCloud(
    scale: Double,
    phase: BreathPhase,
    modifier: Modifier = Modifier,
    /** 0 at the start of a session, 0.55 at the end. Everything that glows
     * fades against it, so the screen gets quieter as the child does. */
    dim: Double = 0.0,
    width: Dp = BreathingCloudMetrics.renderedWidth,
) {
    val isAsleep = phase == BreathPhase.Asleep
    // Eyes close on the hold as well as in sleep — the pause at the top of a
    // breath is where a drowsy face settles.
    val eyesClosed = isAsleep || phase == BreathPhase.Hold
    val height = width * (BreathingCloudMetrics.artHeight / BreathingCloudMetrics.artWidth)
    val scaleFactorDp = width / BreathingCloudMetrics.artWidth

    Box(
        modifier = modifier
            .size(width, height)
            .testTag("sleepy-cloud-${phase.name.lowercase()}")
            // Deliberately not `mascot-<mood>`: a second element answering
            // to that identifier would make every celebration lookup in
            // other screens' UI tests ambiguous — mirrors iOS's own
            // deliberate identifier choice.
            .clearAndSetSemantics { contentDescription = "Cloudmoji" },
    ) {
        Canvas(
            modifier = Modifier
                .fillMaxSize()
                .graphicsLayer {
                    // The breath itself. Driven every frame by the caller
                    // (`SleepyCloudScreen`'s own wall-clock loop), so — like
                    // iOS's `TimelineView`-driven scale — it carries no
                    // `animateFloatAsState` of its own.
                    scaleX = scale.toFloat()
                    scaleY = scale.toFloat()
                },
        ) {
            val scaleFactor = this.size.width / BreathingCloudMetrics.artWidth
            val yOffset = (this.size.height - BreathingCloudMetrics.artHeight * scaleFactor) / 2f
            withTransform({
                translate(0f, yOffset)
                scale(scaleFactor, scaleFactor, pivot = Offset.Zero)
            }) {
                drawHalo(dim)
                drawCloudBody()
                drawCloudHighlights()
                drawUndersideShadow()
                drawBlush()
                drawEyes(eyesClosed)
                drawMouth(phase)
            }
        }

        if (isAsleep) {
            ZzzGlyph(x = 101f, y = 14f, fontSizePx = 12f, alpha = 0.8f, scaleFactorDp = scaleFactorDp)
            ZzzGlyph(x = 109f, y = 5f, fontSizePx = 9f, alpha = 0.6f, scaleFactorDp = scaleFactorDp)
        }
    }
}

/** Every number [BreathingCloud] is drawn from. Mirrors iOS
 * `BreathingCloud`'s own static properties. */
object BreathingCloudMetrics {
    /** The web viewBox. Every coordinate in this file's drawing functions is
     * in these units. */
    const val artWidth: Float = 120f
    const val artHeight: Float = 78f

    /** Drawn at 200 x 130 in the prototype, which is this viewBox at
     * 1.667x. */
    val renderedWidth: Dp = 200.dp

    /** Sideways. A landscape phone gives about 400pt of height, and 130 of
     * it for the cloud plus a title, three buttons and a caption overflowed
     * — the cloud came out clipped against the top edge of the screen. */
    val compactRenderedWidth: Dp = 132.dp
}

// MARK: - Drawing (viewBox-local coordinates; the caller has already applied
// the scale + letterbox transform)

/** A quiet crescent-adjacent halo behind the cloud — mirrors iOS's
 * `RadialGradient` `Ellipse`, approximated here with a circular
 * [Brush.radialGradient] filled into the same oval bounds (Compose has no
 * elliptical-gradient primitive either; iOS's own `RadialGradient` is
 * circular too, just clipped by the `Ellipse()` shape it fills). */
private fun DrawScope.drawHalo(dim: Double) {
    val center = Offset(60f, 46f)
    drawOval(
        brush = Brush.radialGradient(
            colors = listOf(Moonlight.copy(alpha = 0.5f), Moonlight.copy(alpha = 0f)),
            center = center,
            radius = 54f,
        ),
        topLeft = Offset(60f - 54f, 46f - 36f),
        size = Size(108f, 72f),
        alpha = (0.35 * (1 - dim)).toFloat().coerceIn(0f, 1f),
    )
}

private fun DrawScope.drawCloudBody() {
    val circles = listOf(
        Triple(30f, 46f, 20f),
        Triple(52f, 36f, 23f),
        Triple(72f, 30f, 26f),
        Triple(94f, 42f, 19f),
        Triple(42f, 44f, 16f),
    )
    for ((cx, cy, r) in circles) drawCircle(color = CloudWhite, radius = r, center = Offset(cx, cy))
    drawRoundRect(
        color = CloudWhite,
        topLeft = Offset(12f, 48f),
        size = Size(96f, 24f),
        cornerRadius = CornerRadius(12f, 12f),
    )
}

private fun DrawScope.drawCloudHighlights() {
    drawCircle(color = CloudHighlight, radius = 12f, center = Offset(72f, 22f))
    drawCircle(color = CloudHighlight.copy(alpha = 0.7f), radius = 8f, center = Offset(50f, 30f))
}

private fun DrawScope.drawUndersideShadow() {
    drawOval(
        color = CloudShadow.copy(alpha = 0.4f),
        topLeft = Offset(60f - 44f, 68f - 6f),
        size = Size(88f, 12f),
    )
}

/** Fixed opacity, unlike [app.cloudmoji.android.ui.mascot.CloudMascot]'s own
 * mood-dependent blush — a sleepy cloud has one steady, quiet expression. */
private fun DrawScope.drawBlush() {
    val color = Blush.copy(alpha = 0.42f)
    for (cx in floatArrayOf(34f, 86f)) {
        drawOval(color = color, topLeft = Offset(cx - 8f, 58f - 4.5f), size = Size(16f, 9f))
    }
}

private fun DrawScope.drawEyes(closed: Boolean) {
    if (closed) {
        val path = Path()
        for (cx in floatArrayOf(46f, 74f)) {
            path.moveTo(cx - 6f, 52f)
            path.quadraticTo(cx, 42f, cx + 6f, 52f)
        }
        drawPath(path = path, color = MascotEyes, style = Stroke(width = 2.2f, cap = StrokeCap.Round))
    } else {
        drawOval(color = MascotEyes, topLeft = Offset(46f - 2.4f, 51f - 2.8f), size = Size(4.8f, 5.6f))
        drawOval(color = MascotEyes, topLeft = Offset(74f - 2.4f, 51f - 2.8f), size = Size(4.8f, 5.6f))
    }
}

private fun DrawScope.drawMouth(phase: BreathPhase) {
    when (phase) {
        BreathPhase.Asleep ->
            drawOval(color = BlushBeaming.copy(alpha = 0.75f), topLeft = Offset(60f - 4f, 62f - 3f), size = Size(8f, 6f))

        BreathPhase.Inhale ->
            // A small round mouth, the way a person taking a breath in has
            // one.
            drawOval(color = BlushBeaming.copy(alpha = 0.7f), topLeft = Offset(60f - 4.5f, 62f - 4f), size = Size(9f, 8f))

        BreathPhase.Hold, BreathPhase.Exhale -> {
            val path = Path().apply {
                moveTo(55f, 61f)
                quadraticTo(60f, 69f, 65f, 61f)
            }
            drawPath(
                path = path,
                color = MascotEyes.copy(alpha = 0.75f),
                style = Stroke(width = 1.6f, cap = StrokeCap.Round),
            )
        }
    }
}

/** One "z", drawn as a real `Text` overlay positioned in the same viewBox
 * units the [Canvas] above uses — mirrors
 * [app.cloudmoji.android.ui.mascot.CloudMascot]'s own sparkle-overlay
 * technique, and the reason: Compose's `DrawScope` has no built-in text
 * primitive as convenient as SwiftUI's `Text` inside a `ZStack`.
 *
 * An approximation, not a pixel match: iOS's `.position(x:y:)` centres the
 * glyph on the given point; `Modifier.offset` here places its top-left
 * corner there instead, since Compose has no equivalent single-call
 * "position the center" for a `Text` sized to its own content without an
 * extra measurement pass. Two small "z"s a few points off-position during
 * the asleep phase is not worth that pass — the same category of trade-off
 * [app.cloudmoji.android.ui.mascot.CloudMascot]'s own sparkle placement
 * already documents. */
@Composable
private fun ZzzGlyph(x: Float, y: Float, fontSizePx: Float, alpha: Float, scaleFactorDp: Dp) {
    val fontSizeDpValue = fontSizePx * scaleFactorDp.value
    Text(
        text = "z",
        color = Moonlight.copy(alpha = alpha),
        fontSize = fontSizeDpValue.sp,
        modifier = Modifier.offset(x = scaleFactorDp * x, y = scaleFactorDp * y),
    )
}
